cropify/src/main/java/io/moyuru/cropify/Cropify.kt

```diff
@@ -1,34 +1,84 @@
 package io.moyuru.cropify
 
+import android.content.Context
+import android.graphics.Bitmap
+import android.net.Uri
 import androidx.compose.foundation.gestures.detectDragGestures
 import androidx.compose.foundation.layout.BoxWithConstraints
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.geometry.Offset
 import androidx.compose.ui.geometry.Rect
 import androidx.compose.ui.geometry.Size
 import androidx.compose.ui.graphics.ImageBitmap
+import androidx.compose.ui.graphics.asAndroidBitmap
+import androidx.compose.ui.graphics.asImageBitmap
 import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.platform.LocalContext
 import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.unit.IntSize
 import androidx.compose.ui.unit.dp
 import java.lang.Float.max
 import java.lang.Float.min
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
 
 @Composable
 fun Cropify(
+  uri: Uri,
   state: CropifyState,
+  onImageCropped: (ImageBitmap) -> Unit,
+  onFailedToLoadImage: (Throwable) -> Unit,
   modifier: Modifier = Modifier,
-  option: CropifyOption = CropifyOption()
+  option: CropifyOption = CropifyOption(),
+) {
+  BoxWithConstraints(modifier = modifier) {
+    val context = LocalContext.current
+    var sampledImageBitmap by remember(uri) { mutableStateOf<SampledImageBitmap?>(null) }
+
+    LaunchedEffect(uri) {
+      try {
+        sampledImageBitmap = loadSampledImageBitmap(context, uri, constraints.run { IntSize(maxWidth, maxHeight) })
+        state.loadedUri = uri
+        state.inSampleSize = requireNotNull(sampledImageBitmap).inSampleSize
+      } catch (t: Throwable) {
+        sampledImageBitmap = null
+        onFailedToLoadImage(t)
+      }
+    }
+
+    if (sampledImageBitmap != null) {
```

Comment:
```suggestion
    sampledImageBitmap?.let {
      Cropify(
         bitmap = it.imageBitmap,
         state = state,
         onImageCropped = onImageCropped,
         option = option,
         modifier = Modifier.matchParentSize()
       )
    }
```

if 조건문으로 이미 sampledImageBitmap 이 null 이 아닌 것을 확인했음에도 requireNotNull을 사용하여 다시 한번 null 검사를 수행하는 것이 불필요하고 가독성을 떨어뜨린다고 생각하여 let 구문을 사용하도록 개선해 보았습니다.

또한 requireNotNull 은 어쨌든 예외 발생 가능성이 있으므로 주의하여 사용해야 한다고 생각했습니다.

Reply:
맞습니다. 단순히 null check만 한다면 정말정말 일어나진 않겠지만, 동시성문제가 발생할 수도 있습니다. 널체크하는 순간 아래 구분에서 갑자기 null이 생길 수 있는 상황이 있습니다. let을 사용하면 해당문제는 명확하게 처리가 되기때문에 괜찮을 것같습니다.
혹시나 이해가 안되시면 예시를 알려 드릴게요.